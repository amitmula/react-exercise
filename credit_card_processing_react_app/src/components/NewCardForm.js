import React, { Component } from "react";
import Form from 'react-bootstrap/Form'
import Button from 'react-bootstrap/Button'
import Alert from 'react-bootstrap/Alert'

export default class AddNewCard extends Component {
    constructor(props) {
        super(props);
        this.state = {
            form : {
                balance: 0
            },
            alertBox: false,
            alertVarient: "primary",
            message : [],
            httpSuccess: true
        }
    }

    setShow(val) {
        this.setState({...this.state, alertBox: val})
    }

    handleSubmit = event => {
        fetch(process.env.REACT_APP_API_ENDPOINT + process.env.REACT_APP_API_CONTEXT_PATH + process.env.REACT_APP_API_ADD_NEW_CARD_PATH, {
            method: 'POST',
            headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(this.state.form)
        })
        .then( response => {
            response.ok ? this.setState(Object.assign({}, this.state, { httpSuccess: true })) : this.setState(Object.assign({}, this.state, { httpSuccess: false }))
            return response.json()
        }).then( resp => {
            if(this.state.httpSuccess) {
                this.setState(Object.assign({}, this.state, { message: resp.message, alertVarient: "success", alertBox: true, form: {balance: 0}}))
            } else {
                this.setState(Object.assign({}, this.state, { message: resp.errors[0].msg, alertVarient: "danger", alertBox: true }))
            }
        }).catch( error => {
            console.log("fetch api error -> ", error)
        });
        event.preventDefault();
    }
    

    handleFormChange(event) {
        var addForm = this.state.form
        addForm[event.target.id] = event.target.value
        this.setState({...this.state, form: addForm})
    }

    render() {
        return (
        <div>
            <Form onSubmit={(e) => this.handleSubmit(e)}>
                <Form.Group controlId="name">
                    <Form.Label>Name</Form.Label>
                    <Form.Control type="text" placeholder="Enter name" value={this.state.form.name ? this.state.form.name : ''} onChange={(e) => this.handleFormChange(e)} required />
                </Form.Group>

                <Form.Group controlId="number">
                    <Form.Label>Card Number</Form.Label>
                    <Form.Control type="text" placeholder="Enter card number" value={this.state.form.number ? this.state.form.number : ''} onChange={(e) => this.handleFormChange(e)} required />
                </Form.Group>

                <Form.Group controlId="limit">
                    <Form.Label>Limit</Form.Label>
                    <Form.Control type="text" placeholder="Enter Limit" value={this.state.form.limit ? this.state.form.limit : ''} onChange={(e) => this.handleFormChange(e)} required />
                </Form.Group>

                <Button variant="primary" type="submit" >
                    Add
                </Button>
            </Form>
            { this.state.alertBox ?  <Alert variant={this.state.alertVarient} className="alertBox" onClose={() => this.setShow(false)} dismissible>
                {this.state.message}
            </Alert> : '' }
        </div>)
    }
}