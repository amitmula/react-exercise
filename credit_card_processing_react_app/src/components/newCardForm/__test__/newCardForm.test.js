import React from 'react'
import ReactDom from 'react-dom'

import NewCardForm from '../newCardForm'

import { render, cleanup } from '@testing-library/react'
import "@testing-library/jest-dom/extend-expect"
import renderer from 'react-test-renderer'

afterEach(cleanup)

it("renders without crashing", () => {
    const div = document.createElement("div")
    ReactDom.render(<NewCardForm />, div)
    ReactDom.unmountComponentAtNode(div)
})

it("renders newCardForm with props", () => {
    const { getByTestId } = render(<NewCardForm testProp="testPropVal" />)
    expect(getByTestId('newCardForm').getAttribute("test-attribute")).toBe('testPropVal');
})

it("renders cardList with props", () => {
    const { getByTestId } = render(<NewCardForm testProp="some test val" />)
    expect(getByTestId('newCardForm').getAttribute("test-attribute")).toBe('some test val');
})

it("matches snapshot", () => {
    const tree = renderer.create(<NewCardForm testProp="testPropVal" />).toJSON()
    expect(tree).toMatchSnapshot()
})
