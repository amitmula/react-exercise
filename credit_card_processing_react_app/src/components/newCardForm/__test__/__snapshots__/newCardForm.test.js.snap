// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<form
  className=""
  data-testid="newCardForm"
  onSubmit={[Function]}
  test-attribute="testPropVal"
>
  <div
    className="form-group"
  >
    <label
      className="form-label"
      htmlFor="name"
    >
      Name
    </label>
    <input
      className="form-control"
      id="name"
      onChange={[Function]}
      placeholder="Enter name"
      required={true}
      type="text"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <label
      className="form-label"
      htmlFor="number"
    >
      Card Number
    </label>
    <input
      className="form-control"
      id="number"
      onChange={[Function]}
      placeholder="Enter card number"
      required={true}
      type="text"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <label
      className="form-label"
      htmlFor="limit"
    >
      Limit
    </label>
    <input
      className="form-control"
      id="limit"
      onChange={[Function]}
      placeholder="Enter Limit"
      required={true}
      type="text"
      value=""
    />
  </div>
  <button
    className="btn btn-primary"
    disabled={false}
    type="submit"
  >
    Add
  </button>
</form>
`;
